include/seqan3/alphabet/adaptation/char.hpp

```diff
@@ -0,0 +1,247 @@
+#endif
+//!\endcond
+
+#include <limits>
+
+#include <meta/meta.hpp>
+
+#include <seqan3/alphabet/concept_pre.hpp>
+#include <seqan3/core/detail/int_types.hpp>
+
+namespace seqan3::detail
+{
+//!\addtogroup adaptation
+//!\{
+
+//!\brief The list of types that are defined as char adaptations.
+using char_adaptations            = meta::list<char,
+                                               char16_t,
+                                               char32_t>;
+//!\brief The corresponding list of rank types
+using char_adaptations_rank_types = meta::list<std::uint_least8_t,
+                                               std::uint_least16_t,
+                                               std::uint_least32_t>;
+
+//!\brief Metafunction that indicates whether a type is a char alphabet adaptation.
+template <typename type>
+struct is_char_adaptation :
+    public std::conditional_t<meta::in<char_adaptations, type>::value, std::true_type, std::false_type>
+{};
+
+//!\brief Shortcut for seqan3::detail::is_char_adaptation.
+template <typename type>
+constexpr bool is_char_adaptation_v = is_char_adaptation<type>::value;
+
+//!\}
+} // namespace seqan3::detail
+
+namespace seqan3
+{
+
+//!\addtogroup adaptation
+//!\{
+
+// ------------------------------------------------------------------
+// type metafunctions
+// ------------------------------------------------------------------
+
+//!\brief Specialisation of seqan3::underlying_char for char types.
+//!\tparam char_type One of `char`, `char16_t` or `char32_t`.
+//!\sa seqan3::underlying_char_t
+template <typename char_type>
+//!\cond
+    requires detail::is_char_adaptation_v<char_type>
+//!\endcond
+struct underlying_char<char_type>
+{
+    //!\brief The same type as char_type.
+    using type = char_type;
+};
+
+//!\brief Specialisation of seqan3::underlying_rank for char types.
+//!\tparam char_type One of `char`, `char16_t` or `char32_t`.
+//!\sa seqan3::underlying_rank_t
+template <typename char_type>
+//!\cond
+    requires detail::is_char_adaptation_v<char_type>
+//!\endcond
+struct underlying_rank<char_type>
+{
+    //!\brief An unsigned integer type of the same size as `char_type`.
+    using type = meta::at<detail::char_adaptations_rank_types,
+                          meta::find_index<detail::char_adaptations, char_type>>;
+};
+
+// ------------------------------------------------------------------
+// value metafunctions
+// ------------------------------------------------------------------
+
+//!\brief Specialisation of seqan3::alphabet_size that delegates for char types.
+//!\tparam char_type One of `char`, `char16_t` or `char32_t`.
+//!\sa seqan3::alphabet_size_v
+template <typename char_type>
+//!\cond
+    requires detail::is_char_adaptation_v<char_type>
+//!\endcond
+struct alphabet_size<char_type>
+{
+    //!\brief Smallest unsigned integral type that can hold value;
+    using type = detail::min_viable_uint_t<static_cast<uint64_t>(std::numeric_limits<char_type>::max()) + 1 -
```

Comment:
the alphabet size is always of the type bigger than char_type and since we have char32_t, we definitely need uint64_t.